File: core/r/R/enrichment_dot.R
# Recipe: dot plot of GSEA / ORA enrichment results, clusterProfiler style.
# Input is an enrichment table; common column names are auto-detected. When an
# NES column is present the plot is a GSEA dot plot (x = NES); otherwise an ORA
# dot plot (x = gene ratio). Dot color encodes the adjusted p, size the gene
# count. Descriptive: no test is run here, the enrichment is computed upstream.

.enrich_columns <- function(df, spec) {
  nm <- names(df)
  pick <- function(cands, override = NULL) {
    if (!is.null(override) && nzchar(override)) return(override)
    hit <- cands[cands %in% nm]; if (length(hit)) hit[1] else NA_character_
  }
  list(
    term  = pick(c("Description", "ID", "pathway", "term", "Term", "NAME"), spec$data$label),
    padj  = pick(c("p.adjust", "padj", "qvalue", "q_value", "FDR", "pvalue", "p_value")),
    count = pick(c("Count", "setSize", "count", "size")),
    ratio = pick(c("GeneRatio", "gene_ratio", "generatio")),
    nes   = pick(c("NES", "nes", "enrichmentScore", "ES")))
}

.parse_ratio <- function(x) {
  vapply(as.character(x), function(s) {
    if (grepl("/", s, fixed = TRUE)) {
      ab <- suppressWarnings(as.numeric(strsplit(s, "/", fixed = TRUE)[[1]]))
      if (length(ab) == 2 && is.finite(ab[2]) && ab[2] != 0) ab[1] / ab[2] else NA_real_
    } else suppressWarnings(as.numeric(s))
  }, numeric(1), USE.NAMES = FALSE)
}

recipe_enrichment_dot <- function(df, spec) {
  cols <- .enrich_columns(df, spec)
  if (is.na(cols$term)) stop("enrichment_dot: no term/Description column found; set it with --label")
  if (is.na(cols$padj)) stop("enrichment_dot: no adjusted-p column found")

  term  <- as.character(df[[cols$term]])
  padj  <- suppressWarnings(as.numeric(df[[cols$padj]]))
  count <- if (!is.na(cols$count)) suppressWarnings(as.numeric(df[[cols$count]])) else rep(NA_real_, length(term))
  gsea_mode <- !is.na(cols$nes)
  if (gsea_mode) {
    xval <- suppressWarnings(as.numeric(df[[cols$nes]])); xlab_default <- "NES"
  } else if (!is.na(cols$ratio)) {
    xval <- .parse_ratio(df[[cols$ratio]]); xlab_default <- "gene ratio"
  } else {
    xval <- -log10(padj); xlab_default <- "-log10 adjusted p"
  }

  keep <- is.finite(padj) & is.finite(xval) & !is.na(term)
  term <- term[keep]; padj <- padj[keep]; count <- count[keep]; xval <- xval[keep]
  if (length(term) < 1) stop("enrichment_dot: no usable rows")
  df_used <- data.frame(term = term, x = xval, p_adjust = padj, count = count, stringsAsFactors = FALSE)

  topn <- spec$appearance$top_n %||% 15
  idx  <- utils::head(order(padj), topn)
  pd <- data.frame(term = term[idx], x = xval[idx], padj = padj[idx],
                   count = count[idx], stringsAsFactors = FALSE)
  pd$term <- factor(pd$term, levels = pd$term[order(pd$x)])
  has_size <- !all(is.na(pd$count))

  p <- ggplot2::ggplot(pd, ggplot2::aes(x = x, y = term))
  p <- if (has_size) {
    p + ggplot2::geom_point(ggplot2::aes(size = count, colour = padj))
  } else {
    p + ggplot2::geom_point(ggplot2::aes(colour = padj), size = 3)
  }
  p <- p +
    ggplot2::scale_colour_gradient(low = "#C1432B", high = "#3B6DB3", name = "p.adjust") +
    ggplot2::scale_size_continuous(name = "count") +
    ggplot2::labs(x = spec$appearance$x_label %||% xlab_default, y = NULL,
                  title = spec$appearance$title) +
    pub_theme(spec$appearance$theme) +
    ggplot2::theme(legend.position = "right", axis.title.y = ggplot2::element_blank())
  if (gsea_mode) {
    p <- p + ggplot2::geom_vline(xintercept = 0, linetype = "dashed",
                                 linewidth = 0.3, colour = "grey50")
  }

  stats_df <- data.frame(term = as.character(pd$term), x = pd$x, p_adjust = pd$padj,
                         count = pd$count, stringsAsFactors = FALSE)
  names(stats_df)[2] <- if (gsea_mode) "NES" else if (!is.na(cols$ratio)) "gene_ratio" else "neg_log10_padj"

  test_meta <- list(name = if (gsea_mode) "GSEA dot plot" else "ORA dot plot",
                    n_terms_total = length(term), n_terms_shown = nrow(pd), x_axis = xlab_default)
  resolved <- list(family = "enrichment_dot",
                   label = if (gsea_mode) "GSEA dot plot" else "ORA dot plot")
  methods <- .enrich_methods(gsea_mode, length(term), nrow(pd), xlab_default)
  build_script <- function(in_name, fig_stub) .enrich_emit_script(in_name, fig_stub, cols, gsea_mode, topn)

  list(plot = p, stats = stats_df, test_meta = test_meta, resolved = resolved,
       methods = methods, build_script = build_script,
       label = slugify(tools::file_path_sans_ext(spec$data$file)),
       df_used = df_used, clean_steps = "No cleaning applied; enrichment table used as-is.",
       width = 7.2, height = max(4.0, 0.34 * nrow(pd) + 1.6))
}

.enrich_methods <- function(gsea_mode, n_total, n_shown, xlab) {
  v    <- pkg_versions("ggplot2")
  rver <- paste(R.version$major, R.version$minor, sep = ".")
  kind <- if (gsea_mode) "GSEA" else "over-representation (ORA)"
  paste0(
    sprintf("The top %d of %d enriched terms from a %s analysis were shown as a dot plot. ",
            n_shown, n_total, kind),
    sprintf("The x-axis is %s, dot color the adjusted p-value, and dot size the gene count. ", xlab),
    "The enrichment analysis was performed upstream; this figure summarises its output. ",
    sprintf("Drawn in R %s with ggplot2 %s.", rver, v$ggplot2))
}

.enrich_emit_script <- function(in_name, fig_stub, cols, gsea_mode, topn) {
  xline <- if (gsea_mode) {
    sprintf('d$x <- as.numeric(raw[["%s"]]); xlab <- "NES"', cols$nes)
  } else if (!is.na(cols$ratio)) {
    sprintf('d$x <- sapply(strsplit(as.character(raw[["%s"]]), "/"), function(a) as.numeric(a[1])/as.numeric(a[2])); xlab <- "gene ratio"', cols$ratio)
  } else {
    sprintf('d$x <- -log10(as.numeric(raw[["%s"]])); xlab <- "-log10 adjusted p"', cols$padj)
  }
  vline <- if (gsea_mode) '  + geom_vline(xintercept = 0, linetype = "dashed")' else ""
  c(
    "#!/usr/bin/env Rscript",
    "# Standalone reproduction. Run inside the pinned pubplot container.",
    "suppressPackageStartupMessages(library(ggplot2))",
    "",
    sprintf('raw <- read.csv("%s", check.names = FALSE)', in_name),
    sprintf('d <- data.frame(term = as.character(raw[["%s"]]), padj = as.numeric(raw[["%s"]]))',
            cols$term, cols$padj),
    if (!is.na(cols$count)) sprintf('d$count <- as.numeric(raw[["%s"]])', cols$count) else "d$count <- NA",
    xline,
    sprintf('d <- d[order(d$padj), ][seq_len(min(%d, nrow(d))), ]', topn),
    "d$term <- factor(d$term, levels = d$term[order(d$x)])",
    'p <- ggplot(d, aes(x, term)) + geom_point(aes(size = count, colour = padj)) +',
    '  scale_colour_gradient(low = "#C1432B", high = "#3B6DB3") + labs(x = xlab, y = NULL) + theme_classic()',
    paste0("p <- p", vline),
    sprintf('ggsave("%s.pdf", p, width = 7.2, height = 6)', fig_stub),
    'cat("Reproduced figure.\\n")'
  )
}

File: core/python/pubready/recipes/enrichment_dot.py
"""Recipe: dot plot of GSEA / ORA enrichment results, clusterProfiler style.
Mirrors the R recipe. An NES column -> GSEA dot plot (x = NES); otherwise an ORA
dot plot (x = gene ratio). Color = adjusted p, size = gene count. Descriptive; no
test is run here, the enrichment is computed upstream."""
import os

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from ..bundle import slugify
from ..theme import apply_pub_style

_TERM = ["Description", "ID", "pathway", "term", "Term", "NAME"]
_PADJ = ["p.adjust", "padj", "qvalue", "q_value", "FDR", "pvalue", "p_value"]
_COUNT = ["Count", "setSize", "count", "size"]
_RATIO = ["GeneRatio", "gene_ratio", "generatio"]
_NES = ["NES", "nes", "enrichmentScore", "ES"]


def _pick(df, cands, override=None):
    if override:
        return override
    for c in cands:
        if c in df.columns:
            return c
    return None


def _parse_ratio(series):
    out = []
    for s in series.astype(str):
        try:
            if "/" in s:
                a, b = s.split("/")[:2]
                out.append(float(a) / float(b))
            else:
                out.append(float(s))
        except Exception:
            out.append(np.nan)
    return np.array(out)


def recipe_enrichment_dot(df, spec):
    term_col = _pick(df, _TERM, spec["data"].get("label"))
    padj_col = _pick(df, _PADJ)
    if term_col is None:
        raise ValueError("enrichment_dot: no term/Description column found; set it with --label")
    if padj_col is None:
        raise ValueError("enrichment_dot: no adjusted-p column found")
    count_col = _pick(df, _COUNT)
    ratio_col = _pick(df, _RATIO)
    nes_col = _pick(df, _NES)

    term = df[term_col].astype(str).to_numpy()
    padj = pd.to_numeric(df[padj_col], errors="coerce").to_numpy()
    count = (pd.to_numeric(df[count_col], errors="coerce").to_numpy()
             if count_col else np.full(len(term), np.nan))
    gsea = nes_col is not None
    if gsea:
        xval = pd.to_numeric(df[nes_col], errors="coerce").to_numpy(); xlab = "NES"
    elif ratio_col:
        xval = _parse_ratio(df[ratio_col]); xlab = "gene ratio"
    else:
        xval = -np.log10(padj); xlab = "-log10 adjusted p"

    keep = np.isfinite(padj) & np.isfinite(xval)
    term, padj, count, xval = term[keep], padj[keep], count[keep], xval[keep]
    if len(term) < 1:
        raise ValueError("enrichment_dot: no usable rows")
    df_used = pd.DataFrame({"term": term, "x": xval, "p_adjust": padj, "count": count})

    ap = spec["appearance"]
    topn = int(ap.get("top_n", 15))
    idx = np.argsort(padj)[:topn]
    t, x, pj, ct = term[idx], xval[idx], padj[idx], count[idx]
    yo = np.argsort(x)
    t, x, pj, ct = t[yo], x[yo], pj[yo], ct[yo]
    has_size = not np.all(np.isnan(ct))

    w = 7.2
    h = max(4.0, 0.34 * len(t) + 1.6)
    fig, ax = plt.subplots(figsize=(w, h))
    ypos = np.arange(len(t))
    if has_size:
        rng = np.nanmax(ct) - np.nanmin(ct)
        sizes = 30 + 200 * ((ct - np.nanmin(ct)) / rng if rng else np.zeros_like(ct))
    else:
        sizes = 70
    sc = ax.scatter(x, ypos, c=pj, s=sizes, cmap="coolwarm_r", edgecolors="none")
    ax.set_yticks(ypos); ax.set_yticklabels(t, fontsize=9)
    ax.set_xlabel(ap.get("x_label") or xlab)
    if gsea:
        ax.axvline(0, ls="--", lw=0.6, color="grey")
    if ap.get("title"):
        ax.set_title(ap["title"], fontweight="bold")
    apply_pub_style(ax, ap.get("theme"))
    fig.colorbar(sc, ax=ax, label="p.adjust", fraction=0.046, pad=0.04)
    if has_size:
        handles, labels = sc.legend_elements(prop="sizes", num=3, alpha=0.6,
                                             func=lambda s: (s - 30) / 200 * rng + np.nanmin(ct))
        ax.legend(handles, labels, title="count", loc="lower right", frameon=False, fontsize=8)
    fig.tight_layout()

    stats_df = pd.DataFrame({
        "term": t, ("NES" if gsea else "gene_ratio" if ratio_col else "neg_log10_padj"): x,
        "p_adjust": pj, "count": ct})
    test_meta = {"name": "GSEA dot plot" if gsea else "ORA dot plot",
                 "n_terms_total": len(term), "n_terms_shown": len(t), "x_axis": xlab}
    methods = _methods(gsea, len(term), len(t), xlab)
    return {
        "fig": fig, "stats": stats_df, "test_meta": test_meta, "methods": methods,
        "df_used": df_used, "clean_steps": ["No cleaning applied; enrichment table used as-is."],
        "label": slugify(os.path.splitext(spec["data"]["file"])[0]),
        "width": w, "height": h,
        "build_script": lambda in_name, fig_stub: _script(in_name, fig_stub, term_col, padj_col, count_col, ratio_col, nes_col, topn),
    }


def _methods(gsea, n_total, n_shown, xlab):
    from importlib.metadata import version as v
    import platform
    kind = "GSEA" if gsea else "over-representation (ORA)"
    return (
        f"The top {n_shown} of {n_total} enriched terms from a {kind} analysis were shown as a "
        f"dot plot. The x-axis is {xlab}, dot color the adjusted p-value, and dot size the gene "
        "count. The enrichment analysis was performed upstream; this figure summarises its output. "
        f"Drawn in Python {platform.python_version()} with matplotlib {v('matplotlib')}."
    )


def _script(in_name, fig_stub, term_col, padj_col, count_col, ratio_col, nes_col, topn):
    if nes_col:
        xline = f'x = pd.to_numeric(raw["{nes_col}"], errors="coerce").to_numpy(); xlab = "NES"'
    elif ratio_col:
        xline = (f'x = raw["{ratio_col}"].astype(str).map(lambda s: float(s.split("/")[0])/float(s.split("/")[1]) '
                 f'if "/" in s else float(s)).to_numpy(); xlab = "gene ratio"')
    else:
        xline = f'x = -np.log10(pd.to_numeric(raw["{padj_col}"], errors="coerce").to_numpy()); xlab = "-log10 adjusted p"'
    count_line = (f'ct = pd.to_numeric(raw["{count_col}"], errors="coerce").to_numpy()'
                  if count_col else "ct = np.full(len(raw), 60.0)")
    return f'''#!/usr/bin/env python3
# Standalone reproduction. Run inside the pinned pubready container.
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np, pandas as pd

raw = pd.read_csv("{in_name}")
term = raw["{term_col}"].astype(str).to_numpy()
padj = pd.to_numeric(raw["{padj_col}"], errors="coerce").to_numpy()
{count_line}
{xline}
idx = np.argsort(padj)[:{topn}]
t, x, pj, ct = term[idx], x[idx], padj[idx], ct[idx]
yo = np.argsort(x); t, x, pj, ct = t[yo], x[yo], pj[yo], ct[yo]
fig, ax = plt.subplots(figsize=(7.2, max(4.0, 0.34*len(t)+1.6)))
sc = ax.scatter(x, np.arange(len(t)), c=pj, s=70, cmap="coolwarm_r", edgecolors="none")
ax.set_yticks(np.arange(len(t))); ax.set_yticklabels(t)
ax.set_xlabel(xlab)
for s in ("top", "right"): ax.spines[s].set_visible(False)
fig.colorbar(sc, ax=ax, label="p.adjust", fraction=0.046, pad=0.04)
fig.tight_layout(); fig.savefig("{fig_stub}.pdf")
print("Reproduced figure.")
'''
